pkg/appregistry/builder.go

```diff
@@ -0,0 +1,222 @@
+package appregistry
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"database/sql"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+
+	"k8s.io/klog"
+
+	"github.com/operator-framework/operator-registry/pkg/apprclient"
+	"github.com/operator-framework/operator-registry/pkg/sqlite"
+)
+
+type RegistryImageBuilder interface {
+	Build() error
+}
+
+type ImageAppender interface {
+	Append(from, to, layer string) error
+}
+
+type ImageAppendFunc func(from, to, layer string) error
+
+func (f ImageAppendFunc) Append(from, to, layer string) error {
+	return f(from, to, layer)
+}
+
+type AppregistryImageBuilder struct {
+	Appender ImageAppender
+
+	// options
+	From, To            string
+	AuthToken           string
+	AppRegistryEndpoint string
+	AppRegistryOrg      string
+	DatabasePath        string
+	CacheDir            string
+
+	// derived
+	CleanOutput bool
+	ManifestDir string
+	DatabaseDir string
+}
+
+func NewAppregistryImageBuilder(config *AppregistryBuildOptions, options ...AppregistryBuildOption) (*AppregistryImageBuilder, error) {
```

Comment:
Taking the config be a formal parameter defeats at least one reason for using the variadic options pattern -- that is, always requiring the config argument (i.e. `NewAppregistryImageBuilder(nil)` vs. `NewAppregistryImageBuilder()`). 

Also, a common configuration can always be stored as a slice of pre-baked option closures.

Reply:
I did this so that I can directly use the options built by cobra flags (which is the style `oc` uses).

I'll update with a compromise: a method on `AppregistryBuildOptions`, `ToOption` that builds an `AppregistryBuildOption` from an existing option struct. Then the initializer can only have variadic options.